src/jobserver-posix.cc

```diff
@@ -0,0 +1,170 @@
+  while (flag_char_ < strlen(makeflags)) {
+    while (flag_char_ < strlen(makeflags) &&
+           !isblank(static_cast<unsigned char>(makeflags[flag_char_]))) {
+      flag_.push_back(static_cast<unsigned char>(makeflags[flag_char_]));
+      flag_char_++;
+    }
+
+    if (!flag_.empty())
+      flags_.push_back(flag_);
+
+    flag_.clear();
+    flag_char_++;
+  }
+
+  // --jobserver-auth=<val>
+  for (size_t n = 0; n < flags_.size(); n++)
+    if (flags_[n].find(AUTH_KEY) == 0)
+      flag_ = flags_[n].substr(strlen(AUTH_KEY));
+
+  // --jobserver-fds=<val>
+  if (flag_.empty())
+    for (size_t n = 0; n < flags_.size(); n++)
+      if (flags_[n].find(FDS_KEY) == 0)
+        flag_ = flags_[n].substr(strlen(FDS_KEY));
+
+  // -j 1
+  if (flag_.empty())
+    for (size_t n = 0; n < flags_.size(); n++)
+      if (flags_[n].find("-j") == 0)
+        jobserver_closed_ = true;
+
+  jobserver_name_.assign(flag_);
+  const char* jobserver = jobserver_name_.c_str();
+
+  // Check for fifo type first.
+  if (jobserver_name_.find(FIFO_KEY) == 0)
+    jobserver_fifo_ = true;
+
+  // Warn if jobserver type is unknown (neither fifo nor pipe).
+  if (!jobserver_fifo_ && sscanf(jobserver, "%d,%d", &rfd_, &wfd_) != 2)
+    if (!jobserver_name_.empty())
+      Warning("invalid jobserver value: '%s'", jobserver);
+
+  // Open FDs to the pipe if needed.
+  if (jobserver_fifo_) {
+    rfd_ = open(jobserver + strlen(FIFO_KEY), O_RDONLY | O_NONBLOCK);
+    wfd_ = open(jobserver + strlen(FIFO_KEY), O_WRONLY);
+  }
+
+  // Exit on failure to open FDs, build non-parallel for invalid passed FDs.
+  if (rfd_ >= 0 && wfd_ >= 0)
+    Info("using jobserver: %s", jobserver);
+  else if (jobserver_fifo_ && (rfd_ == -1 || wfd_ == -1))
+    Fatal("failed to open jobserver: %s: %s", jobserver, strerror(errno));
+  else
+    jobserver_closed_ = true;
+
+  // Fallback to non-parallel if jobserver-capable parent has no jobserver.
+  if (jobserver_closed_)
+    rfd_ = wfd_ = 0;
+
+  // Signal that we have initialized but do not have a token yet.
+  token_count_ = -1;
+}
+
+Jobserver::~Jobserver() {
+  Clear();
+
+  if (!jobserver_closed_ && rfd_ >= 0)
+    close(rfd_);
+  if (!jobserver_closed_ && wfd_ >= 0)
+    close(wfd_);
+
+  rfd_ = -1;
+  wfd_ = -1;
+}
+
+bool Jobserver::Enabled() const {
+  return rfd_ >= 0 && wfd_ >= 0;
+}
+
+bool Jobserver::Acquire() {
+  // The first token is implicitly handed to a process.
+  if (token_count_ <= 0) {
+    token_count_ = 1;
+    return true;
+  }
+
+  // Fallback to non-parallel building if pipe is closed.
+  if (jobserver_closed_) {
+    return false;
+  }
+
+  char token;
+
+  int ret = read(rfd_, &token, 1);
+  if (ret < 0 && errno != EAGAIN && errno != EWOULDBLOCK) {
+    jobserver_closed_ = true;
+    if (!jobserver_fifo_)
+      Warning("pipe closed: %d (mark the command as recursive)", rfd_);
+    else
+      Fatal("failed to read from jobserver: %d: %s", rfd_, strerror(errno));
```

Comment:
Is there a way to fallback gracefully to the usual mode when this happens instead?

In general, it is better to avoid calling Fatal() in methods like these, because these conditions cannot be properly unit-tested.

Reply:
the remaining calls to Fatal() are for the following cases:

1. MAKEFLAGS from environment told ninja that a fifo object needs to have file descriptors open for it. The open() syscall is ran and errored. 
2. File descriptors have been opened to the fifo object, but reading from it has failed not due to blocking. This likely means that the fifo object has been deleted while we have file descriptors that point to nothing, or maybe a problem with the filesystem itself.
3. File descriptors have been opened to the fifo object, but writing to it has failed, again suggesting that the fifo object has been deleted while we have file descriptors that point to nothing or a problem with the filesystem.

They can become warnings, but I think they protect execution from continuing when something is extremely wrong. It should be exceedingly rare for the program to step into these Fatal() calls and they are probably out of scope for unit tests anyway, but I'll let you decide.